upf/upf_pfcp_api.c

```diff
@@ -1669,7 +1702,43 @@ handle_update_far (upf_session_t * sx, pfcp_update_far_t * update_far,
 	    ip_udp_gtpu_rewrite (&update->forward, fib_index, is_ip4);
 	  }
 	//TODO: transport_level_marking
-	//TODO: forwarding_policy
+	//forwarding_policy  >> oln: Implementation
+	if (ISSET_BIT (far->update_forwarding_parameters.grp.fields,
+		       UPDATE_FORWARDING_PARAMETERS_FORWARDING_POLICY))
+	  {
+	    policy_id =
+	      far->update_forwarding_parameters.forwarding_policy.identifier;
+	    hash_ptr = hash_get_mem (gtm->forwarding_policy_by_id, policy_id);
+	    if (hash_ptr)
+	      {
+		if (update->forward.fp_pool_index != ~0)	// old policy is preconfigured
+		  {
+		    fp_entry = pool_elt_at_index (gtm->upf_forwarding_policies, update->forward.fp_pool_index);	// old policy
+		    fp_entry->ref_cnt--;	// decrement FAR reference counter in old policy table
```

Comment:
this is too early. The policy is still in active use by the session at this point.

Reply:
Thanks for noticing it. I moved reference decremental to `pfcp_update_apply`
Also I added decremental to `pfcp_disable_sesssion`